src/Tools/ExternalAccess/AspNetCore/AddPackage/AspNetCoreAddPackageCodeAction.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImport;
+using Microsoft.CodeAnalysis.AddPackage;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.AddPackage;
+
+/// <inheritdoc cref="InstallPackageData"/>
+/// <param name="packageNamespaceName">The fully qualified name of the namespace that should be added as a
+/// <c>using/Import</c> in the file if not already present. Should be of the form <c>A.B.C.D</c> only.</param>
+internal readonly struct AspNetCoreInstallPackageData(string packageSource, string packageName, string packageVersionOpt, string packageNamespaceName)
+{
+    public readonly string PackageSource = packageSource;
+    public readonly string PackageName = packageName;
+    public readonly string PackageVersionOpt = packageVersionOpt;
+    public readonly string PackageNamespaceName = packageNamespaceName;
+}
+
+internal static class AspNetCoreAddPackageCodeAction
+{
+    /// <summary>
+    /// Try to create the top level 'Add Nuget Package' code action to add to a lightbulb.
+    /// </summary>
+    /// <param name="document">The document the fix is being offered in.</param>
+    /// <param name="position">The location where the fix is offered.  This will also influence where the using/Import
+    /// directive is added.</param>
+    /// <param name="installPackageData">Information about the package to be added.</param>
+    public static async Task<CodeAction?> TryCreateCodeActionAsync(
+        Document document,
+        int position,
+        AspNetCoreInstallPackageData installPackageData,
+        CancellationToken cancellationToken)
+    {
+        var textChanges = await GetTextChangesAsync(
+            document, position, installPackageData.PackageNamespaceName, cancellationToken).ConfigureAwait(false);
```

Comment:
computing text changes (to add the using-directive) is unfun/hard for asp to do.  All we require from them is the name of the namespace, and where the fix is being applied, and we'll figure out what text changes are needed (which may be nothing if the using directive is already there).